Addition of two music management options to the QUICK ACCES menu.

ZFEbHVUE opened the pull request:
Hi  all.

I have added TWO OPTIONS IN THE “QUICK ACCESS” MENU To keep track of the music you like when you hear it in a favorite music directory in /userdata/favorite_songs/ with a file called favorite.txt. This file contains the paths and names of the songs that will be played randomly.

The default music folder are:    /usr/shar/batocera/music or /userdata/music/
The favorite music folder called favorite_songs will be created in  /userdata/

UNDER EMULATIONSTATION:

Two Options are have been created :

SAVE TO FAVORITE MUSIC 
USE FAVORITE MUSIC DIRECTORY    with a graphic switch  (no /yes) (it appears only when the first song is saved into favorite folder)

You can selected what you want to use favorite or default folder to play your songs randomly.

A copy of music already present in the favorite folder will not be copied.

Three files have been modified: 

/es-app/src/guis/GuiMenu.cpp
/es-core/src/guis/AudioManager.h
/es-core/src/guis/AudioManager.cpp

All following displayed texts have been translated into all present languages in /locale/lang/

"SAVE TO FAVORITE MUSIC"
"USE FAVORITE MUSIC DIRECTORY"
"Song added to favorite list!"
 "Error: Could not open favorites file."
"No song is currently playing."
"Favorite music directory activated!"
"Default music directory activated!"
"This song is already in Favorites."  

I already make many batocera  compilation with my fork and it works with 42-dev.

My last compilation: OS version: 42-dev-4cc7ed936c 2025/03/02 04:10

Enjoy !

ZFEbHVUE.

ZFEbHVUE commented:
Thank Fabrice for you recommendations.
I see what I can do in this your way

Have a nice day.

Stéphane.

ZFEbHVUE commented:
Bonjour Fabrice.

J'ai résolu les conflits.

J'ai donc utlisé Utils::FileSystem::readAllLines la méthode loadFavoriteSongs ( dans AudioManager.cpp)

J'ai utilisé des class de paths (dans paths.h et paths.cpp)  pour linux et Windows. Maintenant, j'ai recompilé le tout sous batocera et ça fonctionne bien; mais je n'ai aucuns moyens de tester sous Windows. je suis sous linux.

Je n'ai pas mis d'icones devant mes options...  

J'ai un problème de traduction quand c'est de l'arabe ou de l'hebreu; mes traductions ne commencent pas à droite pour "SKIP TO THE NEXT SONG et pour "SAVE TO FAVORITE MUSIC".

fabricecaruso commented:
- Il y a toujours un chemin hardcodé  "/userdata/favorite_music/favorites.txt"
- Je pense que tu n'as pas bien compris comment marchait la classe Paths, dont l'interet est de pouvoir injecter facilement selon les forks d'ES a partir du contructeur de la classe, et permettre un override via fichier de configuration (via loadCustomConfiguration) : Regarde comment ca marche pour getMusicPath() avec mMusicPath et calque toi sur ce modele.

ZFEbHVUE commented:
Bonsoir Fabrice 

Merci pour tes explications et ton temps accordé. je vais essayer d'y remédier.

ZFEbHVUE/Stéphane.

fabricecaruso commented:
Tell us when it's finished. When it's finished, please squash your code into a single commit, please, so we can evaluate.
Also : Why do you use a custom folder for favorites, can't you just make it simple with a "favorites.m3u" file in the existing user music folder ?

ZFEbHVUE commented:
Fabrice.
J'ai terminé ce sur quoi je travaillais. Vous pouvez tester. J'ai essayé de faire comme tu m'avais dit en éliminant les chemins harcodés en dur. Peut-être que ce n'est pas ce que tu voulais; tu,me diras.  J'ai fixé les conflits que j'avais eu avec tous les fichiers de language. tous les textes que j'ai ajouté sont traduit dans tous les langues.

Ps Désolé, j'ai quelques soucis de compilation c'est derniers temps avec batocera.

fabricecaruso commented:
J'avoue que c'est très très brouillon tout ca, et je vais répondre en francais vu que tu sembles être francais aussi, ca sera plus simple.
Je ne comprends pas le sens de tout ce que tu cherches a faire en terme de "chemins" :

- getFavoriteMusicPath n'est utilisé nulle part donc inutile. En meme temps les favoris sont forcement "utilisateur" donc je vois pas l'interet de getFavoriteMusicPath / mFavoriteMusicPath. Avec Batocera, on aurait pas le droit d'écrire dans ce dossier de toutes facons....
- getUserFavoriteMusicPath : A quoi ca sert ? Juste a s'assurer que le repertoire existe ? Aucun interet. ou alors c'est plutot getFavoritesMusicFilePath() qui est inutile, vu que le nom de fichier est hardcodé...
- getWin32UserDataPath() et getUserDataPath() : c'est utilisé nulle part... Pourquoi c'est encore là ?

J'avoue que je ne vois pas l'interet de tout ces chemins - et surtout créer un nouveau dossier juste pour stocker 1 seul fichier texte. 
Tout cela porte vraiment a confusion. Tu dois absolument remettre ca au clair et faire simple.

Mon avis est qu'il est préferable de stocker ton fichier dans le dossier getUserMusicPath() / mUserMusicPath, là ou sont les musiques ajoutées par les utilisateurs. Ca sera plus simple niveau code, mais aussi pour les utilisateurs. Mais dans ce cas -> rien de ce que tu as ajouté dans Paths n'a d'interet.

Aussi, l'idéal serait de faire simple : un fichier m3u ( c'est très basique comme format ), comme ca les utilisateurs peuvent le bidouiller avec des outils genre VLC. Mais bon, si tu n'y arrives pas, un fichier txt fera l'affaire ( mais le m3u serait vraiment top !!! ).

Pour terminer : il n'y a rien de prévu pour retirer un element des favoris....

ZFEbHVUE commented:
Fabrice 
I made what you asked. New definition of paths. Creation of a favorite.m3u file into /userdata/music/

It works like:

1) In Quick Access, you get “SKIP TO THE NEXT SONG” with (“NOW PLAYING”), which lets you see what music is currently playing. Just below this, you get a new option called “SAVE TO FAVORITES” with (“Save current song to favorites playlist”).

2) when you want to save the music you're listening at the first time, you go to “SAVE TO FAVORITES” and a file will be created called favorite.m3u in /userdata/music/ and immediately a new option appears on the screen called “USE FAVORITES PLAYLIST” with a switch in the off position. 

3 If you continue to listen to music and return to music you've already saved, then “SAVE TO FAVORITES” with (“Save current song to favorites playlist”) disappears and is replaced by “ERASE SONG FROM FAVORITES” with (“Remove current song from the favorites playlist”) so that you can delete the favorite.m3u file if you wish.

4) you can of course use your favorite file if “USE FAVORITES PLAYLIST” is present with the switch in position, but to make it available, you need to press “BACK”. Then, when you return to Quick Access, the music played comes from your favorite .m3u file. Now you no longer have “SAVE TO FAVORITES” but only “ERASE SONG FROM FAVORITES”. If all your favorite music is deleted, favorite.m3u will be erased and you'll see the “SAVE TO FAVORITES” present and “USE FAVORITES PLAYLIST” menus disappear, and you return automaticaly into the default music folder. 

5) If, for example, you have music in the file favorite. m3u and “USE FAVORITES PLAYLIST” is “on” and you play your favorite music randomly, you can return to the default music folder  by setting the “USE FAVORITES PLAYLIST” switch to “off”, so after pressing ‘BACK’ and returning to “Quick access” you'll get “SKIP TO THE NEXT SONG” with (“NOW PLAYING”), but the “ERASE SONG FROM FAVORITES” menu will disappear to get “SAVE TO FAVORITES” again with (“Save current song to favourites playlist”) and the “USE FAVORITES PLAYLIST” menu set to “off”.


I need to put again all languages for all menus because before I had some test 

The menus

"SAVE TO FAVORITES"

"Save current song to favorites playlist"

"ERASE SONG FROM FAVORITES"

"Remove current song from the favorites playlist"

"USE FAVORITES PLAYLIST"


The informations tests

"This song is already in favorites."

"Could not open favorites file."

"Song added to favorites!"

"Song removed, playlist is now empty."

"Could not open favorites file for rewriting."

"Song removed from favorites."


Some of them aren't use anymore... 


Sorry for the delay. first overall, it was a personal projects after talking to lbrpdx, he convinced me to make a PR. it takes time to me because ti not a developper and coder. we can talk under discord if you really have time otherwise don't worry I'll always follow your advice

Stéphane (ZFEbHVUE)

fabricecaruso commented:
Please : 
- Squash all this into a single commit : I won't merge a PR with **156 commits** !
- Cancel every changes you made to Paths.cpp and Paths.h : you still have getWin32UserDataPath and getUserDataPath useless methods
I'll review it once done.

ZFEbHVUE commented:
Hello Fabrice.

- I squashed my commits. I put back the original Paths.cpp and Paths.h.
- I created a function in es-core/src/ FavoriteMusicManager {.cpp and .h}.
- I've removed alerts and information messages I did before ( This song is already in favorites / Could not open favorites file. / Song added to favorites! / Song removed, playlist is now empty. / Could not open favorites file for rewriting. / Song removed from favorites. ) 

I've kept only the essentials for the menus. 

SAVE TO FAVORITES" (I kept "iconSound" icon) 
 "Save current song to favorites playlist"

- "ERASE SONG FROM FAVORITES"  (I kept "iconSound" icon).
 "Remove current song from the favorites playlist"

- "USE FAVORITES PLAYLIST" (with a no/yes switch).  (without icon)

- Initially, only "SKIP TO THE NEXT SONG" and "SAVE TO FAVORITES" are present. 
- As soon as the first song is saved, the “USE FAVORITES PLAYLIST” menu appears, with the switch set to “no”. 
- A file /userdata/music/favorites.m3u is created with the path and name of the music.
- The user can keep the default directory to continue saving music in favorites.m3u. 
- If musics is already presents, “ERASE SONG FROM FAVORITES” will replace “SAVE TO FAVORITES”. 
- The user can choose to use favorite playlist by turning on the “USE FAVORITES PLAYLIST” switch. To become active, He need to press “BACK”. Return to “QUICK ACCESS”. Only favorite songs will be played randomly, and only the “ERASE SONG FROM FAVORITES” menu will be available. 
- If the playlist becomes empty, the /userdata/music/favorites.m3u file will be deleted automatically, “USE FAVORITES PLAYLIST” will disappear, “SAVE TO FAVORITES” will replace “ERASE SONG FROM FAVORITES” and the music will be played back in the default music folder.

I've translated the menus in all of them (but there are problems I didn't have before).

Have good week. 

Stéphane.

ZFEbHVUE commented:
Hi Fabrice.

I put an icon "iconSound" in front of “USE FAVORITES PLAYLIST” using the function with addWithDescription but keeping the switch active. So I added a description “Only play music from your saved favorites”.  I might not add descriptions to all the menus I've created, to lighten the visual side and reduce the need for language translations.

fabricecaruso commented:
Hi,

This PR is not mergeable in the current state.

Check the "Files changed" tab in the PR : There are 8 commits, and 64 changed files. Most of changed files just include the code from the master you merged manually. ( you should have rebased instead ).

I won't test and merge this as long as the PR is in this state : It could contain only the files you changed.
